### PR TITLE
14.0 pricelist tax

### PR DESCRIPTION
**Pull request by bguillot**
This modules alllows you to set a pricelist in tax included.

It will automatically replace the tax excluded tax by the included tax on the sale order line.

It is usefull if you want to set a fixed price tax included (for instance : 9.99€) on the pricelist for B2C and an untaxed fixed price on the pricelist of B2B

### CODE REVIEW

File: sale_order_pricelist_tax/models/tax.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            else:
+                mtax[tax.company_id.id or 0][tax.amount] = tamount[tax.amount]
+            prev_cpny = cpny
+        return mtax
+
+    def _get_substitute_taxes(self, record, taxes, map_tax):
+        # TODO shortify this code
+        if record._name == "sale.order.line":
+            cpny = record.order_id.company_id.id or self.company_id.id or 0
+        elif record._name == "account.invoice.line":
```
Comment (legalsylvain) [inline sale_order_pricelist_tax/models/tax.py]: AFAIK, ``account.invoice.line`` doesn't exist in V14.

File: sale_order_pricelist_tax/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2018 Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Pricelist Tax",
+    "version": "14.0.1.0.0",
+    "category": "Sale",
+    "author": "Akretion, Odoo Community Association (OCA)",
+    "depends": [
+        "sale",
+        "l10n_fr",
```
Comment (legalsylvain) [inline sale_order_pricelist_tax/__manifest__.py]: please don't depend on localization, only for demo data and test.

**Review by legalsylvain — COMMENTED** [review]
Hi. Quick review. Some remarks inline. 

I'm wondering if we shouldn't cut the code in two module, an "accounting" part that would work for the account.move.line, and a module that would depend on it, for the sales.order.line.

Don't you think ?

File: sale_order_pricelist_tax/models/product.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        string="Price include taxes",
+        default=True,
+        help="If checked, prices of the list are taken account tax include.\n"
+        "We can only update this setting if there is no price item.",
+    )
+
+    @api.constrains("price_include_taxes")
+    def _constrains_pricelist_price_include(self):
+        for rec in self:
+            items = self.env["product.pricelist.item"].search(
```
Comment (rousseldenis) [inline sale_order_pricelist_tax/models/product.py]: Do the search before loop

File: sale_order_pricelist_tax/models/product.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+class ProductPricelistItem(models.Model):
+    _inherit = "product.pricelist.item"
+
+    @api.constrains("base")
+    def _constrains_price_item_price_include(self):
+        for rec in self:
+            price_incl_tax = rec.pricelist_id.price_include_taxes
+            if rec.base in ("pricelist", "standard") and price_incl_tax:
+                raise UserError(
```
Comment (rousseldenis) [inline sale_order_pricelist_tax/models/product.py]: you should raise ValidationError in a constraint

File: sale_order_pricelist_tax/models/product.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    )
+
+    @api.constrains("price_include_taxes")
+    def _constrains_pricelist_price_include(self):
+        for rec in self:
+            items = self.env["product.pricelist.item"].search(
+                [("pricelist_id", "=", rec.id), ("base", "!=", "list_price")]
+            )
+            if items:
+                raise UserError(
```
Comment (rousseldenis) [inline sale_order_pricelist_tax/models/product.py]: You should raise ValidationError in a constraint

File: sale_order_pricelist_tax/models/tax.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            cpny = record.invoice_id.company_id.id or self.company_id.id or 0
+        else:
+            raise ("No other model supported than sale and invoice")
```
Comment (rousseldenis) [inline sale_order_pricelist_tax/models/tax.py]: Please raise an Odoo Exception

**Review by rousseldenis — DISMISSED** [review]
Code review. Some changes

File: sale_order_pricelist_tax/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                line.order_id.fiscal_position_id
+                or line.order_id.partner_id.property_account_position_id
+            )
+            line.tax_id = (
+                fpos.map_tax(taxes, line.product_id, line.order_id.partner_shipping_id)
+                if fpos
+                else taxes
+            )
+
+    def _get_substitute_taxes(self, taxes, map_tax):
```
Comment (sebastienbeau) [inline sale_order_pricelist_tax/models/sale.py]: @bguillot deadcode ? The method already exist on account.tax

File: sale_order_pricelist_tax/models/tax.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            else:
+                mtax[tax.company_id.id or 0][tax.amount] = tamount[tax.amount]
+            prev_cpny = cpny
+        return mtax
+
+    def _get_substitute_taxes(self, record, taxes, map_tax):
+        # TODO shortify this code
+        if record._name == "sale.order.line":
+            cpny = record.order_id.company_id.id or self.company_id.id or 0
+        elif record._name == "account.move.line":
```
Comment (sebastienbeau) [inline sale_order_pricelist_tax/models/tax.py]: We do not have implemented the support on account.move for now, right ?
(If we want to support on account.move we need to depend on account_invoice_pricelist : https://github.com/OCA/account-invoicing/pull/850)

**Comment by gurneyalex** [comment]
@bguillot @sebastienbeau  I'd like to get this one merged. Could you check what is wrong with the CI?

**Comment by sebastienbeau** [comment]
@gurneyalex I am going to work on it right now

**Comment by sebastienbeau** [comment]
@rousseldenis @gurneyalex It's should be ready !

**Comment by sebastienbeau** [comment]
ping @rousseldenis @gurneyalex

**Comment by sebastienbeau** [comment]
@rousseldenis can you update your review to remove the blocking review ?
Thanks

**Comment by rousseldenis** [comment]
@sebastienbeau Could you rebase ?

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -50,11 +50,18 @@ jobs:
             include: "sale_procurement_group_by_commitment_date"
             name: test with OCB
           - container: ghcr.io/oca/oca-ci/py3.6-odoo14.0:latest
-            exclude: "sale_order_line_sequence,sale_procurement_group_by_commitment_date"
+            include: "sale_global_discount"
             makepot: "true"
             name: test with Odoo
           - container: ghcr.io/oca/oca-ci/py3.6-ocb14.0:latest
-            exclude: "sale_order_line_sequence,sale_procurement_group_by_commitment_date"
+            include: "sale_global_discount"
```
Comment (simahawk) [inline .github/workflows/test.yml:57]: do we really need this exception?

**Review by simahawk — COMMENTED** [review]
This PR got stuck on merging...

This is not a migration or yes? I cannot understand from the title

File: .github/workflows/test.yml (hunk 9 above)
Comment (rousseldenis) [reply .github/workflows/test.yml:57]: @bguillot Could you answer this ?

Comment (bguillot) [reply .github/workflows/test.yml:57]: it seems that there is a bug in sale_global_discount when having several inclueded taxes : https://github.com/OCA/sale-workflow/issues/2123

**Comment by rousseldenis** [comment]
@sbidoul Something stuck here

**Comment by sbidoul** [comment]
@rousseldenis there is a failing test.

**Pull request closed without merging**